text/3127-trim-path.md

```diff
@@ -0,0 +1,192 @@
+
+```
+could not initialize thread_rng: /home/username/.cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.8.3/src/rngs/thread.rs
+/home/username/.cargo/registry/src/github.com-1ecc6299db9ec823/rand_chacha-0.3.0/src/guts.rsdescription() is deprecated; use Display
+/home/username/.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.2/src/util_libc.rs
+```
+
+This is undesirable for the following reasons:
+
+1. **Privacy**. `release` binaries may be distributed, and anyone could then see the builder's local OS account username.
+   Additionally, some CI (such as [GitLab CI](https://docs.gitlab.com/runner/best_practice/#build-directory)) checks out the repo under a path where
+   it may include things that really aren't meant to be public. Without sanitising the path by default, this may be inadvertently leaked.
+2. **Build reproducibility**. We would like to make it easier to reproduce binary equivalent builds. While it is not required to maintain
+   reproducibility across different environments, removing environment-sensitive information from the build will increase tolerance on the inevitable
+   environment differences when trying to verify builds.
+
+## Handling sysroot paths
+At the moment, paths to the source files of standard and core libraries, even when they are present, always begin with a virtual prefix in the form
+of `/rustc/[SHA1 hash]/library`. This is not an issue when the source files are not present (i.e. when `rust-src` component is not installed), but
+when a user installs `rust-src` they expect the path to their local copy of source files to be visible. Hence the user should be given an option for
+the local paths to show up in panic messages and backtraces.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`trim-path` is a profile setting which can be set to either `true` or `false`. This is enabled by default when you do a release build,
+such as via `cargo build --release`. You can also manually override it by specifying this option in `Cargo.toml`:
+```toml
+[profile.dev]
+trim-path = true
+
+[profile.release]
+trim-path = false
+```
+
+With `trim-path` option enabled, the compilation process will not introduce any absolute paths into the build output. Instead, paths containing
+certain prefixes will be replaced with something stable by the following rules:
+
+1. Path to the source files of the standard and core library will begin with `/rustc/[rustc version]`.
+   E.g. `/home/username/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs` -> 
+   `/rustc/1.52.1/library/core/src/result.rs`
+2. Path to the working directory will be replaced with `.`. E.g. `/home/username/crate/src/lib.rs` -> `./src/lib.rs`.
+3. Path to packages outside of the working directory will be replaced with `[package name]-[version]`. E.g. `/home/username/deps/foo/src/lib.rs` -> `foo-0.1.0/src/lib.rs`
```

Comment:
This somewhat assumes uniqueness of a crate name + version, but it's possible to have two crates with the same name and version being included in the build from different registries, or git repos, or path sources (though cargo has issues with some overlaps).

```
foo v0.1.0 (/tmp/tmp.NPvWf4JOsU/foo)
├── futures-core v0.3.15 (/tmp/tmp.NPvWf4JOsU/bar)
└── futures-core v0.3.15
```

One option would be to use the "hashed index url" as a leading directory, so in this case assuming `foo` is from crates.io it would be `github.com-1ecc6299db9ec823/foo-0.1.0/src/lib.rs`. Git dependencies similarly have a "hashed git url" that could be used, e.g. `futures-rs-b0bea7d4c3745ece` for `https://github.com/rust-lang/futures-rs`.

Path dependencies I'm not sure about, it could be possible to generate a similar hash based on the actual path, but while that would alleviate privacy issues it would still have reproducibility problems. It's (currently) not possible to have two path dependencies with overlapping name+version, so maybe just a leading segment such as `path/` to distinguish them from non-path dependencies would work?

```
error: package collision in the lockfile: packages bar v0.1.0 (/tmp/tmp.NPvWf4JOsU/bar) and bar v0.1.0 (/tmp/tmp.NPvWf4JOsU/bar2) are different, but only one can be written to lockfile unambiguously
```

---

Alternatively, this could be declared as a non-issue since it will probably never actually occur, but that should be documented so when it _does_ happen to someone they have a better chance of figuring out what's happening.